### PR TITLE
Remove removed items from maps and lists on apply

### PR DESCRIPTION
**Pull request by jennybuckley**
this fixes the broken test "apply_twice_remove"

Still room to refactor within remove.go because there was some code duplication

### CODE REVIEW

File: merge/update.go [hunk 1]
```diff
@@ -149,3 +151,19 @@ func (s *Updater) Apply(liveObject, configObject typed.TypedValue, version field
 	}
 	return newObject, managers, nil
 }
+
+func (s *Updater) removeRemovals(merged, applied typed.TypedValue, lastSet *fieldpath.VersionedSet) (typed.TypedValue, error) {
```
Comment (jennybuckley) [inline merge/update.go]: This is probably not the final function name

File: merge/update.go [hunk 2]
```diff
@@ -149,3 +151,15 @@ func (s *Updater) Apply(liveObject, configObject typed.TypedValue, version field
...
+func (s *Updater) removeDisownedItems(merged, applied typed.TypedValue, lastSet *fieldpath.VersionedSet) (typed.TypedValue, error) {
+	convertedApplied, err := s.Converter.Convert(applied, lastSet.APIVersion)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert applied config to last applied version: %v", err)
+	}
+	appliedSet, err := convertedApplied.ToFieldSet()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create field set from applied config in last applied version: %v", err)
+	}
+	return merged.RemoveItems(lastSet.Set.Difference(appliedSet)), nil
```
Comment (apelisse) [inline merge/update.go]: All of this is surprisingly simpler than I expected :-)

File: typed/deduced.go [hunk 3]
```diff
@@ -130,7 +130,7 @@ func added(lhs, rhs value.Value, path fieldpath.Path, set *fieldpath.Set) {
 	} else if lhs.MapValue == nil && rhs.MapValue != nil {
 		// From leaf to map, add leaf fields of map.
 		fieldsetDeduced(rhs, path, set)
-	} else if lhs.MapValue != nil  && rhs.MapValue == nil {
+	} else if lhs.MapValue != nil && rhs.MapValue == nil {
```
Comment (apelisse) [inline typed/deduced.go]: I've got to get my laptop keyboard fixed :-)

File: typed/remove.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+	w := &removingWalker{
+		value:    value,
+		schema:   schema,
+		toRemove: toRemove,
+	}
+	resolveSchema(schema, typeRef, w)
+}
+
+// doLeaf should be called on leaves before descending into children, if there
+// will be a descent. It modifies w.inLeaf.
```
Comment (apelisse) [inline typed/remove.go]: comment is obsolete

File: typed/remove.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	"sigs.k8s.io/structured-merge-diff/value"
+)
+
+type removingWalker struct {
+	value    *value.Value
+	schema   *schema.Schema
+	toRemove *fieldpath.Set
+}
+
+func removeItemsWithSchema(value *value.Value, toRemove *fieldpath.Set, schema *schema.Schema, typeRef schema.TypeRef) {
```
Comment (apelisse) [inline typed/remove.go]: We need test for this function specifically, and then a few more tests on sets to make sure they are removed properly.

**Pull request merged**